transcribe/whisper.py

```diff
@@ -0,0 +1,243 @@
+    model = load_model(options["model_name"])
+
+    whisper_options = options.copy()
+    whisper_options.pop("model_name")
+
+    if "language" not in options:
+        whisper_options["language"] = get_language(file, options["model_name"])
+
+    audio = load_audio(file)
+
+    result = whisper.transcribe(audio=audio, model=model, **whisper_options)
+
+    result["text"] = result["text"].strip()
+    return result
+
+
+def get_language(file, model_name):
+    model = load_model(model_name)
+    silences = get_silences(file)
+    n_mels = 128 if model_name == "large" else 80
+
+    if len(silences) > 0 and int(silences[0]["start_silence"]) == 0:
+        audio_segment = AudioSegment.from_file(file)
+        start_time = silences[0]["end_silence"] * 1000
+        end_time = (30 + silences[0]["end_silence"]) * 1000
+        clip = audio_segment[start_time:end_time]
+        with tempfile.NamedTemporaryFile() as tmp:
+            clip.export(tmp.name, format="wav")
+            audio = load_audio(tmp.name)
+    else:
+        audio = load_audio(file)
+
+    audioclip = whisper.pad_or_trim(audio)
+    mel = whisper.log_mel_spectrogram(audioclip, n_mels=n_mels).to(model.device)
+    _, probs = model.detect_language(mel)
+
+    return max(probs, key=probs.get)
+
+
+def get_silences(file):
+    file = shlex.quote(file)
+    p = subprocess.Popen(
+        "ffmpeg -i {} -af 'volumedetect' -vn -sn -dn -f null /dev/null".format(file),
+        shell=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    ).communicate()
+    detectvolume = p[-1].decode("utf-8")
+    meanvolume = ffmpegcontentparse(detectvolume, "mean_volume")
+    volume = meanvolume - 1 if meanvolume > -37 else -37
+    p2 = subprocess.Popen(
+        "ffmpeg -i {} -af silencedetect=n={}dB:d=0.5 -f null -".format(file, volume),
+        shell=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    ).communicate()
+    startendsilences = []
+    for item in p2[-1].decode("utf-8").split("\n"):
+        if "silence_start" in item:
+            startendsilences.append(
+                {"start_silence": ffmpegcontentparse(item, "silence_start")}
+            )
+        elif "silence_end" in item:
+            startendsilences[-1] = startendsilences[-1] | {
+                "end_silence": ffmpegcontentparse(item.split("|")[0], "silence_end"),
+                "duration": ffmpegcontentparse(item, "silence_duration"),
+            }
+
+    return startendsilences
+
+
+def ffmpegcontentparse(content, field):
+    lines = content.split("\n")
+    correctline = [idx for idx, s in enumerate(lines) if field in s][0]
+    get_value = lines[correctline].split(":")[-1]
+    value_as_float = float(re.sub(r"[^0-9\-.]", "", get_value, 0, re.MULTILINE).strip())
+    return value_as_float
+
+
+@cache
+def load_model(model_name):
+    # cache the response since it takes some time to load
+    device = "cuda" if torch.cuda.is_available() else "cpu"
```

Comment:
This is very important for getting Whisper to use the GPU on `li-rd-7217.stanford.edu`. Without this it simply uses the CPU.